File: sf_core/tests/common/private_key_helper.rs
use std::fs;
use std::path::{Path, PathBuf};
use tempfile::TempDir;

use super::config::Parameters;
use super::file_utils::repo_root;

/// A temporary private key file, automatically cleaned up when dropped.
pub struct TempPrivateKeyFile {
    _temp_dir: TempDir,
    file_path: PathBuf,
}

impl TempPrivateKeyFile {
    fn new(private_key_contents: &[String]) -> Result<Self, String> {
        let temp_dir = TempDir::new().map_err(|e| format!("Failed to create temp dir: {e}"))?;
        let file_path = temp_dir.path().join("private_key.p8");
        let private_key_str = private_key_contents.join("\n") + "\n";
        fs::write(&file_path, private_key_str)
            .map_err(|e| format!("Failed to write private key file: {e}"))?;

        Ok(Self {
            _temp_dir: temp_dir,
            file_path,
        })
    }

    pub fn path(&self) -> &Path {
        &self.file_path
    }
}

pub fn get_private_key_from_parameters(
    parameters: &Parameters,
) -> Result<TempPrivateKeyFile, String> {
    let private_key_contents = parameters.private_key_contents.as_ref().ok_or_else(|| {
        "SNOWFLAKE_TEST_PRIVATE_KEY_CONTENTS not found in parameters.json".to_string()
    })?;
    TempPrivateKeyFile::new(private_key_contents)
}

fn get_test_private_key_contents() -> Vec<String> {
    let key_path = repo_root()
        .join("tests")
        .join("test_data")
        .join("invalid_rsa_key.p8");

    let key_content = fs::read_to_string(&key_path).unwrap_or_else(|_| {
        panic!(
            "Failed to read test private key file: {}",
            key_path.display()
        )
    });

    key_content.lines().map(|s| s.to_string()).collect()
}

pub fn get_test_private_key_file() -> Result<TempPrivateKeyFile, String> {
    let key_contents = get_test_private_key_contents();
    TempPrivateKeyFile::new(&key_contents)
}

File: python/tests/private_key_helper.py
import tempfile

from pathlib import Path

from .config import get_test_parameters
from .utils import repo_root


# Keep TempPrivateKeyFile alive when a real path is required (old-driver default
# JWT setup and explicit private_key_file tests). New-driver paths prefer PEM via
# get_private_key_pem_from_parameters() so core accepts the string directly.
_temp_key_files_registry = []


class TempPrivateKeyFile:
    """
    File is automatically deleted when this object is garbage collected.
    """

    def __init__(self, private_key_contents: list[str]):
        """Create temporary private key file."""
        self._temp_dir = tempfile.TemporaryDirectory(prefix="snowflake_key_")
        self._file_path = Path(self._temp_dir.name) / "private_key.p8"
        self._file_path.write_text("\n".join(private_key_contents) + "\n")

    def path(self) -> str:
        return str(self._file_path)

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit - cleanup."""
        self.cleanup()

    def __del__(self):
        """Destructor - cleanup when garbage collected."""
        self.cleanup()

    def cleanup(self):
        """Clean up the temporary directory and file."""
        if hasattr(self, "_temp_dir") and self._temp_dir is not None:
            try:
                self._temp_dir.cleanup()
                self._temp_dir = None
            except Exception:
                pass


def get_private_key_pem_from_parameters() -> str:
    """Return private-key PEM text from test parameters.

    Prefers ``SNOWFLAKE_TEST_PRIVATE_KEY_FILE`` when set; otherwise joins
    ``SNOWFLAKE_TEST_PRIVATE_KEY_CONTENTS``. Pass the result as ``private_key``
    to the new driver — no tempfile required.

    Raises:
        RuntimeError: If neither file nor contents are configured.
    """
    test_params = get_test_parameters()

    private_key_file = test_params.get("SNOWFLAKE_TEST_PRIVATE_KEY_FILE")
    if private_key_file:
        return Path(private_key_file).read_text()

    private_key_contents = test_params.get("SNOWFLAKE_TEST_PRIVATE_KEY_CONTENTS")
    if not private_key_contents:
        raise RuntimeError("SNOWFLAKE_TEST_PRIVATE_KEY_CONTENTS not found in test parameters.")

    if isinstance(private_key_contents, list):
        return "\n".join(private_key_contents) + "\n"
    return str(private_key_contents)


def get_private_key_from_parameters() -> str:
    """Get private key file path from test parameters.

    Used by old-driver JWT defaults and tests that specifically exercise
    ``private_key_file``. Prefer :func:`get_private_key_pem_from_parameters`
    for new-driver ``private_key`` string paths.

    Returns:
        Path to the private key file (configured path or a temp file for contents)

    Raises:
        RuntimeError: If SNOWFLAKE_TEST_PRIVATE_KEY_CONTENTS not found
    """
    test_params = get_test_parameters()

    private_key_file = test_params.get("SNOWFLAKE_TEST_PRIVATE_KEY_FILE")
    if private_key_file:
        return private_key_file

    private_key_contents = test_params.get("SNOWFLAKE_TEST_PRIVATE_KEY_CONTENTS")

    if not private_key_contents:
        raise RuntimeError("SNOWFLAKE_TEST_PRIVATE_KEY_CONTENTS not found in test parameters.")

    if not isinstance(private_key_contents, list):
        private_key_contents = str(private_key_contents).splitlines()

    temp_file = TempPrivateKeyFile(private_key_contents)
    # Keep object alive so file persists until program exit
    _temp_key_files_registry.append(temp_file)
    return temp_file.path()


def get_private_key_password() -> str | None:
    test_params = get_test_parameters()
    return test_params.get("SNOWFLAKE_TEST_PRIVATE_KEY_PASSWORD")


def get_test_private_key_path() -> str:
    """Get path to the shared test private key file.

    This key is used for invalid key tests and wiremock tests.

    Returns:
        Path to invalid_rsa_key.p8 in test_data
    """
    return str(repo_root() / "tests" / "test_data" / "invalid_rsa_key.p8")
